crate_universe/extension.bzl

```
@@ -160,16 +167,52 @@ def _crate_impl(module_ctx):
     cargo_bazel = get_cargo_bazel_runner(module_ctx)
     all_repos = []
     for mod in module_ctx.modules:
+        module_annotations = {}
+        repo_specific_annotations = {}
+        for annotation_tag in mod.tags.annotation:
+            annotation_dict = structs.to_dict(annotation_tag)
+            repositories = annotation_dict.pop("repositories")
+            crate = annotation_dict.pop("crate")
+
+            # The crate.annotation function can take in either a list or a bool.
+            # For the tag-based method, because it has type safety, we have to
+            # split it into two parameters.
+            if annotation_dict.pop("gen_all_binaries"):
+                annotation_dict["gen_binaries"] = True
+            annotation = _crate_universe_crate.annotation(**{
+                k: v
+                for k, v in annotation_dict.items()
+                # Tag classes can't take in None, but the function requires None
+                # instead of the empty values in many cases.
+                if v != "" and v != [] and v != {}
+            })
+            if not repositories:
+                _get_or_insert(module_annotations, crate, []).append(annotation)
+            for repo in repositories:
+                _get_or_insert(
+                    _get_or_insert(repo_specific_annotations, repo, {}),
+                    crate,
+                    [],
+                ).append(annotation)
+
         local_repos = []
         for cfg in mod.tags.from_cargo:
             if cfg.name in local_repos:
                 fail("Defined two crate universes with the same name in the same MODULE.bazel file. Use the name tag to give them different names.")
             elif cfg.name in all_repos:
                 fail("Defined two crate universes with the same name in different MODULE.bazel files. Either give one a different name, or use use_extension(isolate=True)")
-            _generate_hub_and_spokes(module_ctx, cargo_bazel, cfg)
+
+            annotations = {k: v for k, v in module_annotations.items()}
+            for crate, values in repo_specific_annotations.get(cfg.name, []).items():
```

```suggestion
            for crate, values in repo_specific_annotations.get(cfg.name, {}).items():
```

I guess this should be a dict, when the key doesn't exist? Otherwise `.items()` fails.